### ISSUE TITLE
An user authentication is failed when a higher directory is specified in base dn for search

### ISSUE THREAD

**Issue by userlocalhost** [StackStorm/st2-auth-backend-ldap#6]
Thank you for providing this great plugin!

BTW, when I change the st2 configuration to use st2-auth-backend-ldap as the authentication backend,
```diff
--- conf/st2.dev.conf.orig      2017-01-11 01:34:16.342011900 +0000
+++ conf/st2.dev.conf   2017-01-12 08:50:46.118023292 +0000
@@ -46,8 +46,17 @@
 logging = st2auth/conf/logging.conf
 
 mode = standalone
-backend = flat_file
-backend_kwargs = {"file_path": "st2auth/conf/htpasswd_dev"}
+backend = ldap
+backend_kwargs = {
+    "ldap_uri": "ldap://localhost",
+    "bind_dn": "***",
+    "bind_pw": "***",
+    "user": { 
+      "base_dn": "OU=hoge,DC=test,DC=local",
+      "search_filter": "(sAMAccountName={username})",
+      "scope": "subtree"
+    }
+  }
 
 # Base URL to the API endpoint excluding the version (e.g. http://myhost.net:9101/)
 api_url = http://127.0.0.1:9101/
```
(some confidential parameters are replaced to asterisk)

`st2 auth` would be success like that.
```
(virtualenv)vagrant@ubuntu2:~$ st2 auth test -p 'test'
+----------+----------------------------------+
| Property | Value                            |
+----------+----------------------------------+
| user     | test                             |
| token    | d60048c1f84a4bd39aaf8288eface664 |
| expiry   | 2017-01-13T08:50:55.269762Z      |
+----------+----------------------------------+
(virtualenv)vagrant@ubuntu2:~$ 
```

But when I change the configuration like this,
```diff
--- conf/st2.dev.conf   2017-01-12 08:50:46.118023292 +0000
+++ conf/st2.dev.conf.new       2017-01-12 09:28:09.766995077 +0000
@@ -52,7 +52,7 @@
     "bind_dn": "***",
     "bind_pw": "***",
     "user": { 
-      "base_dn": "OU=hoge,DC=test,DC=local",
+      "base_dn": "DC=test,DC=local",
       "search_filter": "(sAMAccountName={username})",
       "scope": "subtree"
     }
```

`st2 auth` was failed like this.
```
(virtualenv)vagrant@ubuntu2:~$ st2 auth test -p 'test'
ERROR: 401 Client Error: Unauthorized
MESSAGE: Invalid or missing credentials for url: http://127.0.0.1:9100/tokens

(virtualenv)vagrant@ubuntu2:~$ 
```

And this is the output of `st2auth` when the last authentication request is sent.
```
2017-01-12 09:34:03,731 DEBUG [-] No version specified in URL. Will use default controller.
2017-01-12 09:34:03,738 DEBUG [-] Attempting to fast bind with DN.
2017-01-12 09:34:03,760 DEBUG [-] Connected to LDAP as u:*** 
2017-01-12 09:34:03,761 DEBUG [-] Searching ... DC=test,DC=local 2 (sAMAccountName=test)
2017-01-12 09:34:03,834 DEBUG [-] Failed to uniquely identify the user.
2017-01-12 09:34:03,835 DEBUG [-] LDAP connection closed
2017-01-12 09:34:03,835 AUDIT [-] Invalid credentials provided (remote_addr='127.0.0.1',auth_backend='LDAPAuthenticationBackend')
2017-01-12 09:34:03,837 DEBUG [-] API call failed: None (exception_message='Invalid or missing credentials',exception_data={'comment': None, '_headers': None, 'conditional_response': False, 'detail': 'Invalid or missing credentials', '_app_iter': [''], '_status': '401 Unauthorized', '_headerlist': [('Content-Type', 'text/html; charset=UTF-8'), ('Content-Length', '0')]},exception_class='HTTPUnauthorized')
2017-01-12 09:34:03,838 DEBUG [-] Traceback (most recent call last):
  File "/home/vagrant/st2/virtualenv/local/lib/python2.7/site-packages/pecan/core.py", line 631, in __call__
    self.invoke_controller(controller, args, kwargs, state)
  File "/home/vagrant/st2/virtualenv/local/lib/python2.7/site-packages/pecan/core.py", line 531, in invoke_controller
    result = controller(*args, **kwargs)
  File "/home/vagrant/st2/st2common/st2common/models/api/base.py", line 284, in callfunction
    raise e
HTTPUnauthorized: Invalid or missing credentials

[2017-01-12 09:34:03 +0000] [18317] [DEBUG] Closing connection.
```

I think the later processing should be success because the 'scope' parameter is specified as 'subtree'.

**Comment by nzlosh** [comment]
The message `Failed to uniquely identify the user.` suggests more than one entry matched `sAMAccountName=test` in the search.

It wouldn't be questionable security practice to attempt to authenticate a username against multiple entries as it could lead unauthorised access.  For example, the same account name in different OUs is not necessarily the same person.

**Comment by userlocalhost** [comment]
Thank you for your reply.

It was just as you said.

When I speciify 'DC=test,DC=local' at the search 'base_dn', the extra objects of SearchResultReference appear, like this.
```
vagrant@vagrant:~$ ldapsearch -h localhost -D '***' -w '***' -b 'OU=hoge,DC=test,DC=local' 'CN=Ohyama Hiroyasu'
# extended LDIF
#
# LDAPv3
# base <OU=hoge,DC=test,DC=local> with scope subtree
# filter: CN=Ohyama Hiroyasu
# requesting: ALL
#

# Ohyama Hiroyasu, fuga, hoge, test.local
dn: CN=Ohyama Hiroyasu,OU=fuga,OU=hoge,DC=test,DC=local
objectClass: top
objectClass: person
objectClass: organizationalPerson
objectClass: user
cn: Ohyama Hiroyasu
sn: Ohyama
...
sAMAccountName: oyama-hiroyasu
...

# search result
search: 2
result: 0 Success

# numResponses: 2
# numEntries: 1
vagrant@vagrant:~$ ldapsearch -h localhost -D '***' -w '***' -b 'DC=test,DC=local' 'CN=Ohyama Hiroyasu'
# extended LDIF
#
# LDAPv3
# base <DC=test,DC=local> with scope subtree
# filter: CN=Ohyama Hiroyasu
# requesting: ALL
#

# Ohyama Hiroyasu, fuga, hoge, test.local
dn: CN=Ohyama Hiroyasu,OU=fuga,OU=hoge,DC=test,DC=local
objectClass: top
objectClass: person
objectClass: organizationalPerson
objectClass: user
cn: Ohyama Hiroyasu
sn: Ohyama
...
sAMAccountName: oyama-hiroyasu

# search reference
ref: ldap://ForestDnsZones.test.local/DC=ForestDnsZones,DC=test,DC=local

# search reference
ref: ldap://DomainDnsZones.test.local/DC=DomainDnsZones,DC=test,DC=local

# search reference
ref: ldap://test.local/CN=Configuration,DC=test,DC=local

# search result
search: 2
result: 0 Success

# numResponses: 5
# numEntries: 1
# numReferences: 3
vagrant@vagrant:~$ 
```

And here is the result in case of the problem using `ldap` library directly.
```bash
vagrant@ubuntu:~/prog/python/ldap$ cat test.py 
#!/usr/bin/env python
import ldap

connection = ldap.initialize('ldap://localhost')
connection.simple_bind_s('***', '***')

results = connection.search_s('DC=test,DC=local', ldap.SCOPE_SUBTREE, '(CN=Ohyama Hiroyasu)')

print(len(results))
for r in results:
  print(r)
vagrant@ubuntu:~/prog/python/ldap$ python test.py 
4
('CN=Ohyama Hiroyasu,OU=fuga,OU=hoge,DC=test,DC=local', ...)
(None, ['ldap://ForestDnsZones.test.local/DC=ForestDnsZones,DC=test,DC=local'])
(None, ['ldap://DomainDnsZones.test.local/DC=DomainDnsZones,DC=test,DC=local'])
(None, ['ldap://test.local/CN=Configuration,DC=test,DC=local'])
vagrant@ubuntu:~/prog/python/ldap$ 
```

The reference objects don't be assumed at the implementation and tests.
So, I should be able to make a PR that picks over Entries from the all search results.

**Comment by nzlosh** [comment]
Thanks for the detailed response.  Looking at the results, I'm wondering if we could use a more explicit filter to remove the refernce objects.

Could you try something like `(&(uid={username})(objectClass=user))` to see if you get a unique response to the search query?

**Comment by userlocalhost** [comment]
Thank you for giving advice.

But the results ware same.
```bash
vagrant@ubuntu:~$ cat test.py 
#!/usr/bin/env python
import ldap

connection = ldap.initialize('ldap://ad-server')
connection.simple_bind_s('***', '***')

results = connection.search_s('DC=test,DC=local', ldap.SCOPE_SUBTREE, '(&(CN=Ohyama Hiroyasu)(objectClass=user))')

print(len(results))
for r in results:
  print(r)
vagrant@ubuntu:~$ python test.py 
4
('CN=Ohyama Hiroyasu,OU=fuga,OU=hoge,DC=test,DC=local', {'primaryGroupID': ...
(None, ['ldap://ForestDnsZones.test.local/DC=ForestDnsZones,DC=test,DC=local'])
(None, ['ldap://DomainDnsZones.test.local/DC=DomainDnsZones,DC=test,DC=local'])
(None, ['ldap://test.local/CN=Configuration,DC=test,DC=local'])
vagrant@ubuntu:~$ 
```

Conversely, reference objects also returned when I tried to send a request with a filter that never match any object.
```bash
vagrant@ubuntu:~$ cat test.py 
#!/usr/bin/env python
import ldap

connection = ldap.initialize('ldap://ad-server')
connection.simple_bind_s('***', '***')

results = connection.search_s('DC=test,DC=local', ldap.SCOPE_SUBTREE, '(&(CN=Ohyama Hiroyasu)(hogefoga=puyo))')

print(len(results))
for r in results:
  print(r)
vagrant@ubuntu:~$ python test.py 
3
(None, ['ldap://ForestDnsZones.test.local/DC=ForestDnsZones,DC=test,DC=local'])
(None, ['ldap://DomainDnsZones.test.local/DC=DomainDnsZones,DC=test,DC=local'])
(None, ['ldap://test.local/CN=Configuration,DC=test,DC=local'])
vagrant@ubuntu:~$
```

So I guess that these references are returned in any case under the specific `base_dn`.
(Aside to that, the connecting server is ActiveDirectory on WindowsServer 2013)

**Comment by nzlosh** [comment]
Thanks again for the detailed information.  I've commented in your pull request, but as you rightly point out, the backend doesn't handle references.  This feature would have to be implemented to get a correct response to the initial ldap search.

**Comment by nmaludy** [comment]
I was just bitten with this issue. We were seeing the following error: 
```
2017-04-28 15:48:26,874 57946672 DEBUG ldap_backend [-] Attempting to fast bind with DN.
2017-04-28 15:48:26,891 57946672 DEBUG ldap_backend [-] Connected to LDAP as u:xxx\xxx
2017-04-28 15:48:26,892 57946672 DEBUG ldap_backend [-] Searching ... dc=xxx,dc=xxx,dc=xxx 2 (sAMAccountName=xxx.xxx)
2017-04-28 15:48:26,927 57946672 DEBUG ldap_backend [-] (authenticate) LDAP Error: {'info': '000004DC: LdapErr: DSID-0C090752, comment: In order to perform this operation a successful bind must be completed on the connection., data 0, v2580', 'desc': 'Operations error'} : Type <class 'ldap.OPERATIONS_ERROR'>
2017-04-28 15:48:26,928 57946672 DEBUG ldap_backend [-] LDAP connection closed
2017-04-28 15:48:26,929 57946672 AUDIT handlers [-] Invalid credentials provided (remote_addr='127.0.0.1',auth_backend='LDAPAuthenticationBackend')
2017-04-28 15:48:26,930 57946672 DEBUG hooks [-] API call failed: None (exception_message='Invalid or missing credentials',exception_data={'comment': None, '_headers': None, 'conditional_response': False, 'detail': 'Invalid or missing credentials', '_app_iter': [''], '_status': '401 Unauthorized', '_headerlist': [('Content-Type', 'text/html; charset=UTF-8'), ('Content-Length', '0')]},exception_class='HTTPUnauthorized')
```

When our config had the User base_dn set to a top level path `dc=xxx,dc=xxx,dc=xxx`

We fixed the problem by modifying the user section to the full path of the User's OU: `ou=xxx,ou=xxx,dc=xxx,dc=xxx,dc=xxx`

Is this related?

**Comment by userlocalhost** [comment]
The original reported problem was fixed by #7. So I close this.
